Post excerpt: Add excerpt length control

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Provide a UI control to control the length of the excerpt.
Closes https://github.com/WordPress/gutenberg/issues/34757

Currently there is no way to limit the length of the automatically generated excerpt without code. 
It was requested in discussions in the issue that the feature should limit the length of both the user created and the automatically generated excerpts.

_Description updated January 23._

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To add the ability to control the excerpt length in a design.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The PR adds a new block attribute to the post excerpt block called `excerptLength`. The default value is 55 words, to match the WordPress default length for excerpts.

An excerpt length control is added to the block sidebar. The control uses a `RangeControl` component with the min value of 10 and max value of 100.
The control limits the number of words that are displayed by the post excerpt block on the front.
It does **not** limit the number of words that are in the Excerpt panel in the post tab when editing a post.

The limitation is applied to the display of both auto generated and user-generated excerpts.
In index.php, the number of words is limited by passing `get_the_excerpt` and the `excerptLength` attribute through [wp_trim_words](https://developer.wordpress.org/reference/functions/wp_trim_words/).

In the block editor, outside of a query loop:
- When the block is selected, the full text shows and is editable.
- When it is deselected, the trimmed text shows.

In a query loop, the trimmed text shows.

## Testing Instructions
_Please test with both classic and block themes, and with a theme that filters the excerpt length._

1. Create a new post and add a long paragraph.
2. Add a post excerpt block.
3. Open the block settings sidebar. Confirm that there is a range control combined with a numeric input field, with the label "max number of words".
4. Change the value in the control.
5. Deselect the block and confirm that the trimmed text shows, with the number of words matching the value in the control.
5. Save and view the front, confirm that the excerpt displays the correct number of words.

## Testing instructions for screen reader users
Create a new post.
Add a post excerpt block.
Enter a sample excerpt with more than ten words.
Press tab to move to the settings sidebar.
Move past the Posts tab.
In the Block tab, navigate forwards to the Settings panel.
Navigate past the Show link on new line option.
You should now find a new setting for the text length:
Using VoiceOver, I first hear the number of words, the label, and then the type of control:
"55, Max number of words, slider".
There are three ways to change the value:
- Either use the arrow keys on the slider,
- Tab forwards one step and use the arrow keys in the stepper,
- Tab forwards one step from the slider and enter the number into the field in the stepper.

Change the value to 10.
Navigate back to the editor.
Activate the "read all" command or similar for the screen reader.
Confirm that the excerpt only includes the first ten words.
Save and view the post on the front.
Confirm that the excerpt only includes the first ten words.

carolinan committed:
Post excerpt: Add excerpt length control

carolinan committed:
Update edit.js

carolinan committed:
Update index.php

carolinan committed:
Update core__post-excerpt.json

carolinan committed:
Update index.php

carolinan committed:
Update index.php

skorasaurus commented:
Thanks for the PR; not sure if this is out of the scope of the PR or if it is ready for feedback; but the except_length filter is ignored when it is used on classic PHP themes. 

For example, download and activate the popper theme (it hasn't been updated in a couple years but should work for this sake). 

add the following function and filter to the theme functions.php

```
function new_excerpt_length( $length ) {

	return 10;
}
add_filter( 'excerpt_length', 'new_excerpt_length' );
```

Activate theme. On your homepage, all posts except your newest post should only have an excerpt of 10 words; but when your build of gutenberg is activated, the filter is ignored. 

The excerpt_length filter works as expected when using Gutenberg 14.3.1.

carolinan commented:
Yes, excerpts that are not the block should still be filterable.

carolinan commented:
I was not able to reproduce this. In Popper, with the addition of the filter example code above:
The theme filter is set to 10,
All archives show 10,
Adding an excerpt block to a post and setting the option to 17 words, shows 17 words on the single post, and 10 in the archive where that post is listed.

skorasaurus commented:
> I was not able to reproduce this. In Popper, with the addition of the filter example code above: The theme filter is set to 10, All archives show 10, Adding an excerpt block to a post and setting the option to 17 words, shows 17 words on the single post, and 10 in the archive where that post is listed.

Sorry I was not clear. Yes, the filter is applied when queries built with the query loop and the post excerpts block are used; and the post excerpt block 's excerptLength attribute does not have any value assigned to it; that is great and really helpful! Thank you again for working on this.  

However, for classic themes like popper (and [a custom one that I maintain](https://gitlab.com/cpl/ernesto)) that have a template that is completely php based) (e.g. popper's index.php ) and that don't use the post_excerpt block; the `excerpt_length` php filter is ignored. 

Hopefully, the following video demonstrates the issue that I'm describing: 

https://user-images.githubusercontent.com/955351/196958470-62e06ce6-a7b4-48be-ac3a-7d8671adbf75.mp4

carolinan commented:
But I did use Popper in my test and I got a different result 🤔 -no blocks are used for that query.

carolinan commented:
Regardless of the issue reported above;

The excerpt length control needs to be disabled when a user generated, custom excerpt is used, because the users's text should not be cut off unexpectedly.

1) If I understand correctly, the rest API will return the user generated text if one is added.
To determine if the excerpt is user generated, I tried to compare it with the post content (of course, with the block markup removed, as text only), but this breaks when the excerpt length control is used.

2) I also tried to change the excerpt length to the number of words in the RichText element, onchange, when the user edits it. But this causes problems if the users enters a long text, and then uses the control to set the number of words to something lower, and start editing the text again: The old text is _revealed_ in the rich text element, not replaced.

@spacedmonkey  Would it be reasonable for the rest api to be updated to also return a boolean for wether the excerpt is user generated or not?

carolinan committed:
use wp_trim_words instead of the excerpt length filter

carolinan committed:
Display a warning when the excerpt has a word count that is higher than the set excerpt length.

carolinan commented:
I have updated the PR description, and updated the excerpt length control to limit both user-generated and automatically generated excerpts.
The PR is ready for early testing.

Some concerns:
- I needed to include `@WordPress/wordcount` in `block-library` and I don't think I did it correctly, help wanted.
- I am unsure of if the warning that tells the user that their excerpt contains more words than what the setting allows should be in close proximity to the edited text, or in the sidebar.

carolinan commented on packages/block-library/src/post-excerpt/index.php:
```diff
@@ -18,12 +18,18 @@ function render_block_core_post_excerpt( $attributes, $content, $block ) {
-		return '';
+	/*
+	* The purpose of the excerpt length setting is to limit the length of both
+	* autmatically generated and user-created excerpts.
```
```suggestion
	* automatically generated and user-created excerpts.
```

carolinan committed:
Update packages/block-library/src/post-excerpt/index.php

aristath commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -99,21 +100,36 @@ export default function PostExcerptEditor( {
 	} );
+
+	/*
+	 * translators: If your word count is based on single characters (e.g. East Asian characters),
+	 * enter 'characters_excluding_spaces' or 'characters_including_spaces'. Otherwise, enter 'words'.
+	 * Do not translate into your own language.
+	 */
+	const wordCountType = _x( 'words', 'Word count type. Do not translate!' );
```
Perhaps the `Word count type. Do not translate!` text should include some extra info to convey the message that this should be one of `words`, `characters_excluding_spaces`, or `characters_including_spaces`? Right now nobody will know what to do, unless they search the source code to see the code comment 🤔

carolinan replied:
It is a `translators:` comment not only a code comment, so it will show in the translate.wordpress.org tool, and in Poedit for example.
Correct link 😄 https://developer.wordpress.org/apis/handbook/internationalization/internationalization-guidelines/#descriptions

aristath replied:
Ah you're right.. I missed the `translators:` prefix in the comment. Thanks!

aristath commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -99,21 +100,36 @@ export default function PostExcerptEditor( {
 			value={
-				rawExcerpt ||
-				strippedRenderedExcerpt ||
+				rawExcerpt.trim().split( ' ', excerptLength ).join( ' ' ) ||
+				strippedRenderedExcerpt
+					.trim()
+					.split( ' ', excerptLength )
+					.join( ' ' ) ||
```
Shouldn't this also take into account the value of `wordCountType`, to ensure that it works correctly for `characters_excluding_spaces` and `characters_including_spaces` too? The implementation here makes sense when the value of `wordCountType` is `words`, but we'll need to accommodate the other 2 cases too

carolinan replied:
Working on this 👍

aristath commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -99,21 +100,36 @@ export default function PostExcerptEditor( {
 	) : (
 		<p className={ excerptClassName }>
-			{ strippedRenderedExcerpt || __( 'No post excerpt found' ) }
+			{ strippedRenderedExcerpt
+				.trim()
+				.split( ' ', excerptLength )
+				.join( ' ' ) || __( 'No post excerpt found' ) }
```
Same as the other comment above for using the value of `wordCountType`

carolinan committed:
Support word count types

skorasaurus commented:
Tested as of commit https://github.com/WordPress/gutenberg/pull/44964/commits/86342e2c8bccb579b43bf4065ca19e31315479fc with ernesto theme and [this function commented out ](https://gitlab.com/cpl/ernesto/-/blob/97d2c4cffb92ac6bbdf57acc7676971a1825cc50/includes/theme-loop.php#L8)

_Create a new post, add a long paragraph.
Add a post excerpt block.
Confirm that there is a range control with the text "max number of words in excerpt".
Use the control and confirm that the excerpt displays the correct number of words.
Save and view the front, confirm that the excerpt displays the correct number of words._

Works as expected. 
Note: the post excerpt block's content were empty as indicated in the image below until I saved a draft of the post although I would expect that to be the behavior.

carolinan committed:
Update the word count warning and add speak().

carolinan commented:
In [55edf70](https://github.com/WordPress/gutenberg/pull/44964/commits/55edf7065f50539e8150a55676442c145a9d8825), I changed the "word count exceeded" warning to a more "twitter like" warning:

- When there are 5 words remaining until you reach the excerpt length limit, there is a visible message to the bottom right of the excerpt text, that says for example "5/10".
- When you reach the limit the message says "10/10"
- And when you exceed the limit the message says "-1"  where 1 is the number of words.

Things to solve: 
- Right now, speak() is used to announce the warning to screen readers, but it announces it on every key insert or delete, so it is too noisy.
- Colorize the "-1" warning?

carolinan committed:
Add useDebounce hook to delay speak() while typing an excerpt text.

carolinan commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -16,23 +16,27 @@ import {
 	const isDescendentOfQueryLoop = Number.isFinite( queryId );
 	const userCanEdit = useCanEditEntity( 'postType', postType, postId );
 	const isEditable = userCanEdit && ! isDescendentOfQueryLoop;
+	const debouncedSpeak = useDebounce( speak, 3000 );
```
I only tested this delay by my own ear using VoiceOver, so it might be too long for users who are more used to listening to the screen reader?

aristath replied:
cc @afercia

afercia replied:
Great to see `speak` for this use case 🎉 A few considerations:
- Three seconds seem a bit too much to me. I'd go with 500ms: looking at other occurrences of a debounced `speak` call, that seems the de facto standard. If I remember correctly, 500ms is the same delay used in core in a few places. That said, ideally this value should be adjustable by users. Not all users type at the same speed. Useers who type slowly would benefit from an adjustable delay but this is out of the scope of this PR.
- To my understanding, right now the spoken message occurs at every React render, even when the actual count hasn't changed. That is, the _visual_ counter doesn't change but a new message is announced anyways at any ley press (after the delay). Not sure the mismatch between visual and audible is ideal. Repeating the message each time may be a bit noisy. I'd try to send a message only when `currentWordCount` or `excerptLength` actually change. Maybe `useEffect` may come in handy for that.
- I'd consider to improve the quality of the audible message. As a blind user, hearing messages like:
```
8  10
9  10
10  10
minus 1 
minus 2
minus 3
etc.
```
wouldn't help me that much as the context isn't clear. Ideally, the spoken messages should use specifically crafted, more meaningful, sentences.

carolinan committed:
Merge branch 'trunk' into try/excerpt-length-setting

carolinan committed:
Merge branch 'trunk' into try/excerpt-length-setting

carolinan committed:
Update edit.js

carolinan committed:
Update core__query__deprecated-3.json

annezazu commented:
Just tested this and it's working well, alongside the split settings/styles:

https://user-images.githubusercontent.com/26996883/208972493-c76fde09-d739-497a-9373-236ef47252bc.mov

It was a bit confusing to have the various limits and not go above/below them. I'm not quite sure how we handle that aspect with other blocks. On a design note, "MAX NUMBER OF WORDS IN EXCERPT" seems a bit jarring. Not sure what the options are here.

alexstine commented:
I am really confused as to how to test this. I am not sure why there needs to be 2 controls for the same thing.

carolinan commented:
> I am really confused as to how to test this. I am not sure why there needs to be 2 controls for the same thing.

I did not understand what you meant with there being two controls for the same thing.
Did you mean, that entering custom text is one way to control the length, and adjusting it with the sidebar control is another?

alexstine commented:
@carolinan 

> Did you mean, that entering custom text is one way to control the length, and adjusting it with the sidebar control is another?

Yes.

Just not entirely sure how to test this. Not sure how this PR makes things different.

Thanks.

carolinan commented:
As you know there are two types of excerpts, a custom text entered by the user, and the one automatically generated by WordPress.
It was requested that the feature should limit the length of both types of excerpts, otherwise we can't adjust the excerpts to be the same length on archive pages or in other query loops.

carolinan commented:
@alexstine 
> Not sure how this PR makes things different.

As you know there are two types of excerpts, a custom text entered by the user, and the one automatically generated by WordPress. Currently there is no way to limit the length of the automatically generated excerpt without code.
It was requested that the feature should limit the length of both types of excerpts, otherwise we can't adjust the excerpts to be the same length on archive pages or in other query loops.

The PR adds a new setting to the block settings sidebar and limits the length of the text that is displayed on the front.

I was not able to use the value from the new control to set a hard limit on the entered text, to prevent users from adding more words than allowed by the setting. So instead, when the user enters too many words or is close to entering the max number of words, there is a warning.

This was also made more difficult because there are two places in the interface where you can enter a custom excerpt, in the actual block, and in the post settings sidebar.

carolinan commented:
I personally don't like the capitalized text above the controls, but it is the default and it matches the excerpt length control in the latest post block:
![The block settings sidebar of the latest posts block, with the excerpt length control visible](https://user-images.githubusercontent.com/7422055/210580986-a4cf8a3a-6302-45d1-9543-6a9e72de9899.png)

annezazu commented:
@richtabor if you have time, this would be another good smaller issue to dig into and give feedback on!

richtabor commented:
> If the user edits the excerpt text in the block, a warning is shown if the number of words is higher than the number set in the excerpt length control.

I don't think the need this UI. In today's excerpt field, we don't indicate what will be shown, or would be hidden — I'm not sure we should do the same in the block content area. 

If we want to achieve 1:1 to the front-end, we could have the excerpt shown fully when isSelected, but otherwise not. Show only the specified `excerptLength`, as if it were on the front-end when not selected.  

`Max number of words` looks good to me for the control label. 

<img width="275" alt="CleanShot 2023-01-20 at 12 51 06@2x" src="https://user-images.githubusercontent.com/1813435/213619790-f8a42228-40fb-4e58-a68b-a6646856219b.png">

carolinan commented:
I still think it is a bad experience to let the user customize the text but without indication, not show that customization.

richtabor commented:
> I still think it is a bad experience to let the user customize the text but without indication, not show that customization.

I'd say it's a negative experience _not_ seeing the customized post excerpt — and instead seeing an indicator that some of it is hidden, and therefore uneditable (unless you increase the `excerptLength` value, or edit via the Post sidebar). 

The Post Excerpt block can be used in many contexts — each with differing `excerptLength` values. Just because we see the current post's excerpt, with its `excerptLength` applied, does not mean that value will be the same in any other context, (i.e. any other use of the Post Except block). 

Here's what I'm thinking:

1. Whenever you're dragging/setting the `excerptLength`, show the change. 
2. When you select the RichText component, to change the excerpt, the full excerpt is rendered. 
3. If you deselect the block, the `excerptLength` is applied again, trimming the excerpt. 
4. If the excerpt is longer than the `excerptLenth` the trailing `…` should be applied in the editor when deselected (as it does on the front-end.

richtabor commented:
I also think perhaps we're indexing too heavily on the context of editing a post's excerpt from the block itself, which is unlikely to be within the post content — but rather within the Site Editor, where you are unable to edit the excerpt (at least for now).

carolinan commented:
I did suggest that it should only be available inside a query loop but there was no interest.

richtabor commented:
> I did suggest that it should only be available inside a query loop but there was no interest.

Yea, I feel ya. I think it's fine to have available outside (like other "Theme" blocks), but it is an interesting idea to limit theme blocks from the post editor.

carolinan committed:
Remove the word count message.

carolinan commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -99,21 +108,67 @@ export default function PostExcerptEditor( {
+		trimmedExcerpt = rawOrRenderedExcerpt.trim().split( '', excerptLength );
+	}
+
+	trimmedExcerpt = trimmedExcerpt + '...';
```
I tried using `&hellip`; but it did not work. In the block editor the characters flashed before they were converted and in the site editor only the plain characters showed.

richtabor replied:
Any reason we shouldn't use the `…` directly?

mcsf replied:
I'll ask the same thing Rich asked.

carolinan replied:
No reason, I only missed the question the first time :)

carolinan committed:
Remove the wordcount package

carolinan committed:
Remove the unused CSS

alexstine commented:
If I can be of any more help here, please let me know. I just do not understand how to test this.

carolinan commented:
> If I can be of any more help here, please let me know. I just do not understand how to test this.

@alexstine I am sorry for this confusion.
With the latest changes to the pull request, the excerpt block uses the full length of the excerpt when the block is selected, including in navigation mode. Reading the content in the editor without selecting the blocks will read out the short version.

### Testing instructions
Create a new post.
Add a post excerpt block.
Enter a sample excerpt with more than ten words.
Press tab to move to the settings sidebar.
Move past the Posts tab.
In the Block tab, navigate forwards to the Settings panel.
Navigate past the Show link on new line option.
You should now find a new setting for the text length:
Using VoiceOver, I first hear the number of words, the label, and then the type of control:
"55, Max number of words, slider".
There are three ways to change the value:
- Either use the arrow keys on the slider,
- Tab forwards one step and use the arrow keys in the stepper,
- Tab forwards one step from the slider and enter the number into the field in the stepper.

Change the value to 10.
Navigate back to the editor.
Activate the "read all" command or similar for the screen reader.
Confirm that the excerpt only includes the first ten words.
Save and view the post on the front.
Confirm that the excerpt only includes the first ten words.

carolinan committed:
Merge branch 'trunk' into try/excerpt-length-setting

alexstine commented:
@carolinan Ah, one other detail would have helped a lot. As a keyboard user, I first needed to select another block before I could use virtual navigation to see the truncated words. Sadly, the clash between screen readers and Gutenberg is more complex on Windows than Mac because Gutenberg has different selections, modes, and screen readers have different modes.

After navigating back to the editor, select another block then use a read all command to see the truncated text.

Let me try to code review it now since I know what this is doing.

Thanks for explaining it.

alexstine reviewed (approved):
Code looks clean. Probably would not hurt to get some eyes on all that space splitting magic, I am not the best math person so I feel unqualified to say how this effects word counting.

carolinan commented:
Thank you for testing.

>  I first needed to select another block before I could use virtual navigation to see the truncated words.

That is awkward when there is only one block.

carolinan committed:
Merge branch 'trunk' into try/excerpt-length-setting

carolinan committed:
Merge branch 'trunk' into try/excerpt-length-setting

richtabor commented:
>  I first needed to select another block before I could use virtual navigation to see the truncated words. 

Is there a way we could indicate what's truncated while selected? Like, would wrapping what's truncated in the markup be helpful for a11y? 

If so we could make the wrapped (i.e. truncated) text slightly transparent (maybe 90%) for a visual, and perhaps have assistive text indicating what is truncated, or where the divide exists. Though I'm not 100% that this would work. @alexstine?

alexstine commented:
@richtabor That would be nice but it is likely out of scope for this PR. The whole issue in general is well out of scope of the PR but this is one of Gutenbergs many issues with not communicating contextual info to screen readers. You either have to know exactly that selecting another block, going in to virtual mode, and using down arrow will read the truncated text or we as devs can figure out how to make sure screen reader users do not have less of an experience.

Either way, both of these probably best belong in a follow-up. Do not want to hijack this PR with a ton of extra work.

aristath commented:
I'm reading the comments in this issue and I have a question... Why is this issue blocked? There are small things that can be improved in follow-up PRs, sure... That's always the case with all PRs. But what's actually **blocking** us from adding the excerpt-length control here? 🤔

alexstine commented:
@aristath Nothing from my side. I was simply making a point to say that this should be followed-up on as the experience is less than great for screen reader users.

Thanks.

richtabor commented:
> I was simply making a point to say that this should be followed-up on as the experience is less than great for screen reader users.

Agreed! I just don't want the ideas to fall into the ether :) 

I think this solution is fine, though follow-ups for improved screen reader (and visual) indicators for what's truncated would be good — so you don't have to de-select the block, to understand where the split happens.

carolinan commented:
I have an idea, what if the block had a preview option like the HTML block. With a preview button in the toolbar.

alexstine commented:
@carolinan Good idea. We'll still need to figure out how to make the preview mode accessible at some point, but implementing a base model would be a good start. Once preview is used in multiple places, it should be higher on priority list to improve. It is not unusable currently but it is not very discoverable for users due to some odd issues with tabindex on the block wrapper.

richtabor commented:
> I have an idea, what if the block had a preview option like the HTML block. With a preview button in the toolbar.

I’d rather merge and circle back with something like the previous idea, where you wouldn’t have to do anything to understand what’s truncated.

carolinan commented:
Then, let's click the button 😄

mcsf commented on packages/block-library/src/post-excerpt/index.php:
```diff
@@ -70,3 +76,21 @@ function register_block_core_post_excerpt() {
+ * the excerpt length block setting has no effect.
+ * Returns 100 because 100 is the max length in the setting.
+ */
+if ( is_admin() ||
+	defined( 'REST_REQUEST' ) ||
+	'REST_REQUEST' ) {
```
This is very wide net to cast which will affect all instances of excerpt filtering, regardless of source (Post Excerpt block, Latest Post block, traditional theme templates, etc.) — see #48403.

Why not surround the rendering logic in `render_block_core_post_excerpt` with a pair of `add_filter` / `remove_filter` calls?

mcsf replied:
What I meant by my suggestion was something like this (untested):

```php
function render_block_core_post_excerpt( $attributes, $content, $block ) {
    global $_block_core_post_excerpt_length;

    // ...

    $_block_core_post_excerpt_length = int( $attributes['excerptLength'] );
    add_filter( 'excerpt_length', '_block_core_post_excerpt_filter', PHP_INT_MAX );
    get_the_excerpt();
    remove_filter( 'excerpt_length', '_block_core_post_excerpt_filter' );

    // ...
```

with

```php
function _block_core_post_excerpt_filter() {
  global $_block_core_post_excerpt_length;
  return $_block_core_post_excerpt_length;
}
```

The whole `global` dance is a possible way to work with the fact that `$attributes['excerptLength']` is dynamic.

carolinan replied:
There is already a pull request open that tries to address this.

mcsf replied:
> There is already a pull request open that tries to address this.

Which?

skorasaurus replied:
> > There is already a pull request open that tries to address this.
> 
> Which?

https://github.com/WordPress/gutenberg/pull/48598

tmanhollan replied:
Not sure if it's the one @carolinan was referring to, but I opened #48598 when I discovered that my excerpts outside of the block editor were not obeying my own excerpt_length filter and traced it back to the line you referenced here:

> This is very wide net to cast which will affect all instances of excerpt filtering, regardless of source (Post Excerpt block, Latest Post block, traditional theme templates, etc.) — see #48403.

My [last revision](https://github.com/WordPress/gutenberg/pull/48598/commits/dc82d46dc137b337eaa9433a2c8ff9f3fe7758f3) in my PR did something like what you suggest here:

> Why not surround the rendering logic in `render_block_core_post_excerpt` with a pair of `add_filter` / `remove_filter` calls?

But it only applies to the rendered block on the page; it is not reflected in the editor. The condition on `REST_REQUEST` is the part that makes it work in the editor. I was also concerned that condition casts too wide a net and raised the question in my PR, but received no feedback on that point.

carolinan replied:
Yes, that is the PR I was thinking of

mcsf commented on packages/block-library/src/post-excerpt/index.php:
```diff
@@ -18,12 +18,18 @@ function render_block_core_post_excerpt( $attributes, $content, $block ) {
+	* Because the excerpt_length filter only applies to auto generated excerpts,
+	* wp_trim_words is used instead.
+	*/
+	$excerpt_length = $attributes['excerptLength'];
+	if ( isset( $excerpt_length ) ) {
+		$excerpt = wp_trim_words( get_the_excerpt(), $excerpt_length );
+	} else {
+		$excerpt = get_the_excerpt();
```
Why not inject a filter with the dynamic excerpt length and just call `get_the_excerpt`, which is [already filtered](https://github.com/WordPress/wordpress-develop/blob/b55f01db1dd0c8de654e4175121337fe25f24dca/src/wp-includes/default-filters.php#L200) by `wp_trim_excerpt`?

carolinan replied:
I don't believe I tried that during the development, so there is no obvious reason for not trying that as a possible improvement in a separate PR.

lightbulbheaduk replied:
The addition of the call to "wp_trim_words" will remove any HTML that users have put into a custom excerpt by over-riding get_the_excerpt.  Was this deliberate, or an unfortunate by-product?

mcsf replied:
> The addition of the call to "wp_trim_words" will remove any HTML that users have put into a custom excerpt by over-riding get_the_excerpt. Was this deliberate, or an unfortunate by-product?

I believe that was accidental, so thank you for the flag.

frzsombor replied:
Workaround to bypass the `wp_trim_words` function:

```
function fzs_filter_metadata_registration( $metadata ) {
    if ($metadata['name'] === 'core/post-excerpt') {
	unset($metadata['attributes']['excerptLength']);
    }
    return $metadata;
};
add_filter( 'block_type_metadata', 'fzs_filter_metadata_registration' );
```

lightbulbheaduk replied:
Thank you - that's a really helpful workaround for now

oandregal replied:
This was also reported [in the forums](https://wordpress.org/support/topic/excerpt-stripping-out-html/) and I've confirmed it was introduced in WordPress 6.3. Created an issue at https://github.com/WordPress/gutenberg/issues/66195

mcsf reviewed (commented):
I haven't reviewed `edit.js` much because I'm mainly worried about the PHP issues, which have introduced some bugs here and have correctly triggered test failures in large multisites like WP.com. I think that is the top priority.

That said, the block experience inside the editor also needs to be reviewed more closely. For instance, just from a cursory read and a quick test:

* Two Undo levels are introduced due to the `setAttributes( … ); setExcerpt( … );` sequence of calls. See `__unstableMarkAutomaticChange` for a likely fix.
* As I select in and out of the block, the block will seemingly fail at times, returning "No post excerpt found".
* Ellipsis are appended even when there is no trimming needed.
* The interaction between the `excerptLength` attribute and the fact that a user can edit the excerpt directly is strange.

mcsf commented:
A very important piece of history is https://github.com/WordPress/gutenberg/pull/33366, which **removed** the `wordCount` feature due to conflicts with pre-existing excerpt filters. I had completely forgotten about that PR.

So the obligatory question is: have we worked around the issues that applied to #33366 in 2021? How? Or are they no longer applicable?

carolinan commented:
The PR does not use the rendered content like the previous version, only the excerpt.

carolinan commented:
> As I select in and out of the block, the block will seemingly fail at times, returning "No post excerpt found".

I don't believe that is introduced in this PR. I keep seeing this on WP trunk without Gutenberg (pre 6.2) on different installs (different hosts so different setups). I can't reproduce it consistenly which is why I have not personally created an issue for it. I know this is not the technical answer you are looking for but as a user I experience it as if the excerpt is not being loaded fast enough.


>The interaction between the excerptLength attribute and the fact that a user can edit the excerpt directly is strange.

It is, and that is why there has been so many iterations, and why I have been asking for the option to only be applied to the excerpt when it is not editable.
